go/v1beta1/storage/migration/migration.go

```diff
@@ -0,0 +1,240 @@
+	}
+
+	settingsResponse := map[string]ESSettingsResponse{}
+	if err := DecodeResponse(res.Body, &settingsResponse); err != nil {
+		return fmt.Errorf("error decoding settings response: %s", err)
+	}
+
+	blocks := settingsResponse[indexInfo.Index].Settings.Index.Blocks
+
+	if !(blocks != nil && blocks.Write == "true") {
+		log.Info("Placing write block on index")
+		res, err = e.client.Indices.AddBlock([]string{indexInfo.Index}, "write", e.client.Indices.AddBlock.WithContext(ctx))
+		if err := getErrorFromESResponse(res, err); err != nil {
+			return fmt.Errorf("error placing write block on index: %s", err)
+		}
+
+		blockResponse := &ESBlockResponse{}
+		if err := DecodeResponse(res.Body, blockResponse); err != nil {
+			return fmt.Errorf("error decoding write block response: %s", err)
+		}
+
+		if !(blockResponse.Acknowledged && blockResponse.ShardsAcknowledged) {
+			log.Error("Write block unsuccessful", zap.Any("response", blockResponse))
+			return fmt.Errorf("unable to block writes for index: %s", indexInfo.Index)
+		}
+	}
+
+	newIndexName := e.indexManager.IncrementIndexVersion(indexInfo.Index)
+	err = e.indexManager.CreateIndex(ctx, &IndexInfo{
+		Index:        newIndexName,
+		Alias:        indexInfo.Alias,
+		DocumentKind: indexInfo.DocumentKind,
+	}, true)
+	if err != nil {
+		return fmt.Errorf("error creating target index: %s", err)
+	}
+
+	reindexReq := &ESReindex{
+		Conflicts:   "proceed",
+		Source:      &ReindexFields{Index: indexInfo.Index},
+		Destination: &ReindexFields{Index: newIndexName, OpType: "create"},
+	}
+	reindexBody, _ := EncodeRequest(reindexReq)
+	log.Info("Starting reindex")
+	res, err = e.client.Reindex(
+		reindexBody,
+		e.client.Reindex.WithContext(ctx),
+		e.client.Reindex.WithWaitForCompletion(false))
+	if err := getErrorFromESResponse(res, err); err != nil {
+		return fmt.Errorf("error initiating reindex: %s", err)
+	}
+	taskCreationResponse := &ESTaskCreationResponse{}
+
+	if err := DecodeResponse(res.Body, taskCreationResponse); err != nil {
+		return fmt.Errorf("error decoding reindex response: %s", err)
+	}
+	log.Info("Reindex started", zap.String("taskId", taskCreationResponse.Task))
+
+	reindexCompleted := false
+	pollingAttempts := 10
+	for i := 0; i < pollingAttempts; i++ {
+		log.Info("Polling task API", zap.String("taskId", taskCreationResponse.Task))
+		res, err = e.client.Tasks.Get(taskCreationResponse.Task, e.client.Tasks.Get.WithContext(ctx))
+		if err := getErrorFromESResponse(res, err); err != nil {
+			log.Warn("error getting task status", zap.Error(err))
+			continue
+		}
+
+		task := &ESTask{}
+		if err := DecodeResponse(res.Body, task); err != nil {
+			log.Warn("error decoding task response", zap.Error(err))
+			continue
+		}
+
+		if task.Completed {
+			reindexCompleted = true
+			log.Info("Reindex completed")
+
+			break
+		}
+
+		log.Info("Task incomplete, waiting before polling again", zap.String("taskId", taskCreationResponse.Task))
+		timeSleep(time.Second * 10)
```

Comment:
Initially, we arbitrarily picked how long to wait and for how many attempts. 

To try to prove that out more, we did some load testing. The reindex operation takes documents in [batches of 1000](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-reindex.html#docs-reindex-throttle) so there's not a performance difference until an index has tens of thousands of documents: 

| # occurrences | run 1           | run 2           | run 3           |
|---------------|-----------------|-----------------|-----------------|
| 10            | `10.372776324s` | `10.390366305s` | `10.449805181s` |
| 100           | `10.369810367s` | `10.436173865s` | `10.386952243s` |
| 1000          | `10.392087759s` | `10.452114159s` | `10.457958686s` |
| 10,000        | `10.531893704s` | `10.421936475s` | `10.436905281s` |
| 100,000       | `20.535902343s` | `20.458385589s` | `20.463723459s` |

We ran these tests against Docker Desktop for Mac with a cumulative 6 cores and 8GB of RAM assigned to the Linux VM, which might be more than the typical ES node would have.